File: src/lib/user-block-utils.ts
import type { User } from '@/lib/types';

/** Нормализует `users.blockedUserIds` из Firestore. */
export function normalizeBlockedUserIds(raw: unknown): string[] {
  if (!Array.isArray(raw)) return [];
  return [
    ...new Set(
      raw.filter((x): x is string => typeof x === 'string' && x.trim().length > 0).map((x) => x.trim()),
    ),
  ];
}

/**
 * Личный чат: либо viewer заблокировал partner, либо partner заблокировал viewer
 * (второе известно только если доступен документ partner с полем blockedUserIds).
 */
export function isEitherBlockingFromUserIds(
  viewerId: string,
  viewerBlockedIds: string[] | undefined,
  partnerId: string,
  partnerBlockedIds: string[] | undefined | null,
): boolean {
  if (normalizeBlockedUserIds(viewerBlockedIds).includes(partnerId)) return true;
  if (partnerBlockedIds == null) return false;
  return normalizeBlockedUserIds(partnerBlockedIds).includes(viewerId);
}

export function isEitherBlockingFromUsers(viewer: User, partner: User): boolean {
  return isEitherBlockingFromUserIds(
    viewer.id,
    viewer.blockedUserIds,
    partner.id,
    partner.blockedUserIds ?? null,
  );
}

/** Текст для композера: кто кого заблокировал (если известно). */
export function directChatComposerBlockedHint(
  viewerId: string,
  viewerBlockedIds: string[] | undefined,
  partnerId: string,
  partnerBlockedIds: string[] | undefined | null,
): string {
  if (normalizeBlockedUserIds(viewerBlockedIds).includes(partnerId)) {
    return 'Вы заблокировали этого пользователя. Отправка недоступна — разблокируйте в разделе «Заблокированные» в профиле.';
  }
  if (partnerBlockedIds != null && normalizeBlockedUserIds(partnerBlockedIds).includes(viewerId)) {
    return 'Пользователь ограничил с вами общение. Отправка недоступна.';
  }
  return 'Общение недоступно.';
}

File: mobile/app/lib/features/chat/data/user_block_utils.dart
import '../../../l10n/app_localizations.dart';

// Паритет с web `src/lib/user-block-utils.ts`.

List<String> normalizeBlockedUserIds(Object? raw) {
  if (raw is! List) return const <String>[];
  final out = <String>{};
  for (final e in raw) {
    final s = e.toString().trim();
    if (s.isNotEmpty) out.add(s);
  }
  return out.toList(growable: false);
}

/// Личный чат: viewer заблокировал partner **или** partner заблокировал viewer.
///
/// [partnerBlockedIds] — текущее содержимое `users/{partnerId}.blockedUserIds` (может быть пустым).
/// [partnerUserDocDenied] — `true`, если документ партнёра недоступен (часто partner заблокировал viewer).
bool isEitherBlockingFromUserIds({
  required String viewerId,
  required List<String> viewerBlockedIds,
  required String partnerId,
  required List<String> partnerBlockedIds,
  bool partnerUserDocDenied = false,
}) {
  final my = normalizeBlockedUserIds(viewerBlockedIds);
  if (my.contains(partnerId)) return true;
  if (partnerUserDocDenied) return true;
  return normalizeBlockedUserIds(partnerBlockedIds).contains(viewerId);
}

String directCallBlockedMessageRu({
  required AppLocalizations l10n,
  required String viewerId,
  required List<String> viewerBlockedIds,
  required String partnerId,
  required List<String> partnerBlockedIds,
  bool partnerUserDocDenied = false,
}) {
  if (normalizeBlockedUserIds(viewerBlockedIds).contains(partnerId)) {
    return l10n.block_call_viewer_blocked;
  }
  if (partnerUserDocDenied ||
      normalizeBlockedUserIds(partnerBlockedIds).contains(viewerId)) {
    return l10n.block_call_partner_blocked;
  }
  return l10n.block_call_unavailable;
}
